File: app/src/main/java/com/pomoremote/network/WebSocketClient.java
package com.pomoremote.network;

import android.util.Log;
import com.google.gson.Gson;
import com.pomoremote.timer.TimerState;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.TimeUnit;

public class WebSocketClient {
    private static final String TAG = "WebSocketClient";
    private final OkHttpClient client;
    private WebSocket webSocket;
    private final Gson gson;
    private final WebSocketListener listener;
    private Listener callback;

    public interface Listener {
        void onStateReceived(TimerState state);
        void onConnected();
        void onDisconnected();
    }

    public WebSocketClient(Listener callback) {
        this.callback = callback;
        this.client = new OkHttpClient.Builder()
                .readTimeout(0, TimeUnit.MILLISECONDS)
                .build();
        this.gson = new Gson();
        
        this.listener = new WebSocketListener() {
            @Override
            public void onOpen(@NotNull WebSocket webSocket, @NotNull Response response) {
                Log.d(TAG, "Connected");
                if (WebSocketClient.this.callback != null) {
                    WebSocketClient.this.callback.onConnected();
                }
            }

            @Override
            public void onMessage(@NotNull WebSocket webSocket, @NotNull String text) {
                try {
                    Message msg = gson.fromJson(text, Message.class);
                    if ("state".equals(msg.type) && msg.data != null) {
                        if (WebSocketClient.this.callback != null) {
                            WebSocketClient.this.callback.onStateReceived(msg.data);
                        }
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error parsing message", e);
                }
            }

            @Override
            public void onClosed(@NotNull WebSocket webSocket, int code, @NotNull String reason) {
                Log.d(TAG, "Closed: " + reason);
                if (WebSocketClient.this.callback != null) {
                    WebSocketClient.this.callback.onDisconnected();
                }
            }

            @Override
            public void onFailure(@NotNull WebSocket webSocket, @NotNull Throwable t, @Nullable Response response) {
                Log.e(TAG, "Failure", t);
                if (WebSocketClient.this.callback != null) {
                    WebSocketClient.this.callback.onDisconnected();
                }
            }
        };
    }

    public void connect(String ip, int port) {
        String url = "ws://" + ip + ":" + port + "/ws";
        Request request = new Request.Builder().url(url).build();
        webSocket = client.newWebSocket(request, listener);
    }

    public void close() {
        if (webSocket != null) {
            webSocket.close(1000, "App closed");
        }
    }

    private static class Message {
        String type;
        TimerState data;
    }
}

File: app/src/main/java/com/pomoremote/network/WebSocketClient.kt
package com.pomoremote.network

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.gson.Gson
import com.pomoremote.timer.TimerState
import okhttp3.*
import java.util.concurrent.TimeUnit

class WebSocketClient(private val callback: Listener?) {

    interface Listener {
        fun onStateReceived(state: TimerState)
        fun onConnected()
        fun onDisconnected()
    }

    private val client: OkHttpClient = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .pingInterval(30, TimeUnit.SECONDS)
        .build()

    private var webSocket: WebSocket? = null
    private val gson = Gson()

    private var currentIp: String? = null
    private var currentPort: Int = 0
    @Volatile private var shouldReconnect: Boolean = true
    @Volatile var isConnected: Boolean = false
        private set

    private val reconnectHandler = Handler(Looper.getMainLooper())
    private val reconnectRunnable = object : Runnable {
        override fun run() {
            if (shouldReconnect && !isConnected && currentIp != null) {
                Log.d(TAG, "Attempting reconnect...")
                doConnect()
            }
        }
    }

    private val listener = object : WebSocketListener() {
        override fun onOpen(webSocket: WebSocket, response: Response) {
            Log.d(TAG, "Connected")
            isConnected = true
            reconnectHandler.removeCallbacks(reconnectRunnable)
            callback?.onConnected()
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
            try {
                val msg = gson.fromJson(text, Message::class.java)
                if ("state" == msg.type && msg.data != null) {
                    callback?.onStateReceived(msg.data!!)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error parsing message", e)
            }
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            Log.d(TAG, "Closed: $reason")
            isConnected = false
            callback?.onDisconnected()
            scheduleReconnect()
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            Log.e(TAG, "Failure: ${t.message}")
            isConnected = false
            callback?.onDisconnected()
            scheduleReconnect()
        }
    }

    fun connect(ip: String, port: Int) {
        this.currentIp = ip
        this.currentPort = port
        this.shouldReconnect = true
        doConnect()
    }

    private fun doConnect() {
        webSocket?.cancel()
        val url = "ws://$currentIp:$currentPort/ws"
        val request = Request.Builder().url(url).build()
        webSocket = client.newWebSocket(request, listener)
    }

    private fun scheduleReconnect() {
        if (shouldReconnect) {
            reconnectHandler.removeCallbacks(reconnectRunnable)
            reconnectHandler.postDelayed(reconnectRunnable, RECONNECT_DELAY_MS)
        }
    }

    fun close() {
        shouldReconnect = false
        reconnectHandler.removeCallbacks(reconnectRunnable)
        webSocket?.close(1000, "App closed")
    }

    fun send(message: String) {
        if (webSocket != null && isConnected) {
            webSocket?.send(message)
        }
    }

    private class Message {
        var type: String? = null
        var data: TimerState? = null
    }

    companion object {
        private const val TAG = "WebSocketClient"
        private const val RECONNECT_DELAY_MS = 5000L
    }
}
